File: reference.net/db4o-reference-chapters/src/com/db4odoc/semaphores/LimitLogins.cs
/* Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com */
using System;
using System.IO;
using Db4objects.Db4o;

namespace Db4objects.Db4odoc.Semaphores
{
	/**
	 * This class demonstrates the use of semaphores to limit the
	 * number of logins to a server.
	 */
	public class LimitLogins 
	{
    
		readonly static string HOST = "localhost";
		readonly static int PORT = 4455;
		readonly static string USER = "db4o";
		readonly static string PASSWORD = "db4o";
    
		readonly static int MAXIMUM_USERS = 10; 
    
		public static IObjectContainer Login()
		{
        
			IObjectContainer objectContainer;
			try 
			{
				objectContainer = Db4oFactory.OpenClient(HOST, PORT, USER, PASSWORD);
			} 
			catch (IOException e) 
			{
				return null;
			}
        
			bool allowedToLogin = false;
        
			for (int i = 0; i < MAXIMUM_USERS; i++) 
			{
				if(objectContainer.Ext().SetSemaphore("max_user_check_" + i, 0))
				{
					allowedToLogin = true;
					break;
				}
			}
        
			if(! allowedToLogin)
			{
				objectContainer.Close();
				return null;
			}
        
			return objectContainer;
		}
	}
}

File: reference.vb.net/Db4objects.Db4odoc/Semaphores/LimitLogins.vb
' Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com 
Imports System
Imports System.IO
Imports Db4objects.Db4o

Namespace Db4objects.Db4odoc.Semaphores
    ' 	
    ' 	 This class demonstrates the use of semaphores to limit the
    ' 	 number of logins to a server.
    ' 	
    Public Class LimitLogins

        Shared ReadOnly HOST As String = "localhost"
        Shared ReadOnly PORT As Integer = 4455
        Shared ReadOnly USER As String = "db4o"
        Shared ReadOnly PASSWORD As String = "db4o"

        Shared ReadOnly MAXIMUM_USERS As Integer = 10

        Public Shared Function Login() As IObjectContainer

            Dim objectContainer As IObjectContainer
            Try
                objectContainer = Db4oFactory.OpenClient(HOST, PORT, USER, PASSWORD)
            Catch e As IOException
                Return Nothing
            End Try

            Dim allowedToLogin As Boolean = False

            Dim i As Integer
            For i = 0 To MAXIMUM_USERS - 1 Step i + 1
                If objectContainer.Ext().SetSemaphore("max_user_check_" + i.ToString(), 0) Then
                    allowedToLogin = True
                    Exit For
                End If
            Next

            If Not allowedToLogin Then
                objectContainer.Close()
                Return Nothing
            End If

            Return objectContainer
        End Function
    End Class
End Namespace
